### ISSUE TITLE
Rope fails to rename variable identifier within a generator expression 

### ISSUE THREAD

**Issue by d3r3kk** [python-rope/rope#249]
Issue found with renaming a variable within a generator expression. Exception thrown is:

**RefactoringError: Rename refactoring should be performed on resolvable python identifiers.**

**Python version:** 3.6.6
**OS**: Ubuntu 18.04
**Rope Version**: 0.10.7

## Reproduction Steps
1. Create a folder `project`
2. `cd project`
3. Create a python file called issue.py and copy the following code into it:
````
def do_stuff():
    my_list = [counter + 10 for counter in range(0, 10)]
    print(my_list)

do_stuff()
````
4. Create another python file named renamer.py and copy the following code into it:
````import rope
from rope.base import libutils
from rope.refactor.rename import Rename
import rope.base.project

proj = rope.base.project.Project(
    './',
    save_history=False,
    indent_size=4)

res = libutils.path_to_resource(proj, 'issue.py')

if res.exists:
    renamer = Rename(proj, res, 24)
    changes = renamer.get_changes('valid_count', resources=res)
    for c in changes.changes:
        print(c.new_contents)

````
5. Run the renamer script via the command line `> python renamer.py`

## Expected result
A printout of the `issue.py` file's contents, with the 2 instances of the 'counter` variable renamed to `valid_count`.

## Observed result
Thrown exception:
````
Traceback (most recent call last):
  File "renamer.py", line 14, in <module>
    renamer = Rename(proj, res, 31)
  File "/home/dekeeler/.local/lib/python3.6/site-packages/rope/refactor/rename.py", line 28, in __init__
    'Rename refactoring should be performed'
rope.base.exceptions.RefactoringError: Rename refactoring should be performed on resolvable python identifiers.
````

**Comment by phoenixsampras** [comment]
Hello and help, any workaround for this issue?

**Comment by soupytwist** [comment]
Sorry, I haven't had much time to look into these issues further. If anyone else here is able to pick it up, that would be appreciated.

**Comment by wheelsandmetal** [comment]
Any news on this?

**Comment by mosesontheweb** [comment]
I still encounter this behaviour in v0.16.0. Anyone got tipps? Using it within VSCode.

**Comment by rogedelgado** [comment]
Hello,

I can confirm the issue.I got the error in vim using LanguageCliente-neovim as client. The rope version that I'm using is v0.16.0.

Any advice for this?

**Comment by ArkadyWey** [comment]
Hi, 

I am also still having this issue.

Does anyone have a fix?

**Comment by GreenRobot777** [comment]
Other solution: https://stackoverflow.com/a/59570939/14522024

**Comment by mcepl** [comment]
@lieryan Do you have any idea about this one? We seem to have many very unhappy customers here.

**Comment by lieryan** [comment]
I took a look at this, it seems like there are a number of issues at play here:

- `pyscopes.py::_HoldingScopeFinder` currently assumes that a single logical line of code (i.e. a python statement) can only belong only to a single variable scope, but list/generator expression and lambdas are expressions that can create their own variable scope, so this assumption is incorrect. It may take quite a bit of internal rewrite of `_HoldingScopeFinder` and friends to fix this part of the issue.

 - currently we don't create a variable scope for List/Generator Comprehension, this is in line with Python 2, where Comprehension leaks its loop variables to the surrounding scope, but it's incorrect for Python 3 code

- The loop variable of a List/Generator Comprehension currently aren't considered an assignment statement, which means that unless you have an assignment for the same name elsewhere, rope can't resolve the variable

Probably the quickest way to fix this is to fix that last issue first. That should make rope treat comprehension with Python 2 scope semantics (i.e. leaky loop variable), so renames will work most of the time. In Python 3 though, renames will be a bit overzealous if your loop variable shadows a name in the surrounding scope. We can fix the scoping issue later on.

**Comment by lieryan** [comment]
Closing this ticket as the core renaming issue is already resolved in pull request #318. The slightly incorrect variable scoping issue will be tracked separately in #293.
